File: ui/utilities/video.js
export const getFrameRateFromDashPlayer = (dashPlayer) => {
  const streamInfo = dashPlayer.getActiveStream().getStreamInfo()
  const dashMetrics = dashPlayer.getDashMetrics()
  const dashAdapter = dashPlayer.getDashAdapter()

  const periodIdx = streamInfo.index
  const repSwitch = dashMetrics.getCurrentRepresentationSwitch('video', true)
  const adaptation = dashAdapter.getAdaptationForType(periodIdx, 'video', streamInfo)
  const currentRep = adaptation.Representation_asArray.find((rep) => rep.id === repSwitch.to)

  const frameRateStr = `${currentRep.frameRate}`
  return frameRateFromStr(frameRateStr)
}

export const frameRateFromStr = (frameRateStr) => {
  if (!frameRateStr) return null
  if (frameRateStr.includes('/')) {
    const [numerator, denominator] = frameRateStr.split('/')
    return parseFloat(numerator) / parseFloat(denominator)
  }
  return parseFloat(frameRateStr)
}

export const timecodeFromFrameNumber = (frameNumber, frameRate, showFrames = true) => {
  if (!frameRate) return '00:00;00'
  // Note: this function does not handle hours (more than 59 minutes will display as 60, 61, etc)
  const currentMinutes = `${Math.floor(frameNumber / frameRate / 60)}`.padStart(2, '0')
  const currentSeconds = `${Math.floor(frameNumber / frameRate) % 60}`.padStart(2, '0')
  const currentFrames = `${Math.floor(frameNumber % frameRate)}`.padStart(2, '0')
  let timecode = `${currentMinutes}:${currentSeconds}`
  if (showFrames) {
    timecode = `${timecode};${currentFrames}`
  }
  return timecode
}

export const jumpToPrevFrame = (videoElement, videoFrameRate) =>
  (videoElement.currentTime -= 1 / videoFrameRate)

export const jumpToNextFrame = (videoElement, videoFrameRate) =>
  (videoElement.currentTime += 1 / videoFrameRate)

File: python/utils/video.py
import math
import re

def frame_rate_from_str(frame_rate_str):
    if '/' in frame_rate_str:
        numerator, denominator = map(float, frame_rate_str.split('/'))
        return numerator / denominator
    return float(frame_rate_str)


def timestamp_for_ffmpeg(frame_number, frame_rate):
    total_seconds = frame_number / (frame_rate * 1.0)
    hours = f"{math.floor(total_seconds / 3600):02d}"
    minutes = f"{math.floor(total_seconds / 60) % 60:02d}"
    seconds = f"{math.floor(total_seconds % 60):02d}"
    microseconds = f"{math.floor((total_seconds % 1) * 1000):03d}"
    timestamp = f"{hours}:{minutes}:{seconds}.{microseconds}"
    return timestamp
